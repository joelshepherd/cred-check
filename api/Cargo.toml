[package]
name = "api"
version = "0.1.0"
authors = ["Joel Shepherd <joel@swivid.ink>"]
edition = "2018"

[dependencies]
chrono = { version = "0.4", features = ["serde"] }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
sqlx = { git = "https://github.com/launchbadge/sqlx", features = ["postgres", "macros", "runtime-tokio-rustls", "chrono"] }
tokio = { version = "1.0", features = ["full"] }
warp = "0.3"

[dev-dependencies]
insta = "1.5"
