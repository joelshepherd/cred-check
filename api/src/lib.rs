#[macro_use]
extern crate lazy_static;

mod error;
mod handler;
mod model;
mod route;

use sqlx::{postgres::PgPoolOptions, PgPool};
use std::env;
use warp::Filter;

type Db = PgPool;

lazy_static! {
    static ref SECRET: mwt::SecretKey = mwt::SecretKey::default();
    static ref ENCODER: mwt::Encoder<'static> = mwt::Encoder::new(&SECRET, 8600);
    static ref DECODER: mwt::Decoder<'static> = mwt::Decoder::new(&SECRET, 0);
}

/// Init api
pub async fn init() -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
    let db_config = env::var("DATABASE_URL").expect("Could not find database config.");
    let db = PgPoolOptions::new()
        .connect(&db_config)
        .await
        .expect("Could not connect to database.");

    route::init(db)
}
