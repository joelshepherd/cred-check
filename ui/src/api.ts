// deno-lint-ignore-file camelcase
import { Ok, Err, Result } from "./deps.ts";

// Status code for now
type Error = number;
type ApiResult<T> = Result<T, Error>;

// Auth

interface LoginRequest {
  username: string;
}

interface SignupRequest {
  name: string;
  username: string;
}

interface TokenReply {
  token: string;
}

export function login(input: LoginRequest): Promise<ApiResult<TokenReply>> {
  return request("post", "login", input);
}

export function signup(input: SignupRequest): Promise<ApiResult<TokenReply>> {
  return request("post", "signup", input);
}

// Opinion

interface CreateOpinion {
  source_id: number;
  position: boolean;
  body: string;
}

export interface OpinionReply {
  id: number;
  position: boolean;
  body: string;
}

export function createOpinion(
  input: CreateOpinion
): Promise<ApiResult<OpinionReply>> {
  return request("post", "opinion", input);
}

// Vote
interface CreateVote {
  opinion_id: number;
}

export function createVote(input: CreateVote): Promise<ApiResult<void>> {
  return request("post", "vote", input);
}

// Source

export interface SourceReply {
  id: number;
  title: string;
  canonical_url: string;
  opinions: OpinionReply[];
  votes: [number, number];
}

export function findSource(url: string): Promise<ApiResult<SourceReply>> {
  return request("get", `source/${url}`);
}

export function createSource(url: string): Promise<ApiResult<SourceReply>> {
  return request("post", `source`, { url });
}

async function request<T = unknown>(
  method: "get" | "post",
  path: string,
  data?: unknown
): Promise<ApiResult<T>> {
  const authorization = window.localStorage.getItem("token") ?? "";

  const res = await fetch(`http://localhost:8080/${path}`, {
    method,
    body: data ? JSON.stringify(data) : undefined,
    mode: "cors",
    headers: { authorization, "content-type": "application/json" },
    credentials: "include",
  });

  if (!res.ok) return Err(res.status);

  const reply =
    res.headers.get("content-type") === "application/json"
      ? await res.json()
      : void 0;

  return Ok(reply);
}
