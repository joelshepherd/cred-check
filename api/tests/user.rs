mod tools;

use api::init;
use insta::assert_json_snapshot;
use tools::json_response;
use warp::http::StatusCode;
use warp::test::request;

#[tokio::test]
async fn test_create() {
    let api = init().await;

    let res = request()
        .method("POST")
        .path("/user")
        .body(
            r#"{
                "name": "Tester",
                "username": "tester"
            }"#,
        )
        .reply(&api)
        .await;

    assert_eq!(res.status(), StatusCode::CREATED);
    assert_json_snapshot!(json_response(res));
}
