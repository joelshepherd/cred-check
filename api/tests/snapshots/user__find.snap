---
source: api/tests/user.rs
expression: json_response(res)

---
{
  "id": "[id]",
  "name": "Test"
}
